torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1227,12 +1272,16 @@ void ProcessGroupNCCL::workCleanupLoop() {
         lock,
         std::chrono::milliseconds(kWatchdogThreadSleepMillis),
         [&]() -> bool { return terminateProcessGroup_.load(); });
+    // Bump up heart beat by one in case no collective call.
+    heartbeat_.fetch_add(1);
 
     for (auto it = workMetaList_.begin(); it != workMetaList_.end();
          /* no increment */) {
       auto& work = *it;
       work.checkAndSetException();
       bool timedOut = work.checkTimeout();
+      // Bump up heart beat by one.
+      heartbeat_.fetch_add(1);
```

Comment:
i mean.. i tend to agree with Ke. But i also understand why @fduwjj did it this way.  It is possible we have a large number of work's and each one takes some amount of time to process.  

Basically, all that i think matters is that we are able to tune the heartbeat frequency adequately.  Example:  we expect to have O(100) works and processing each one takes O(10ms) on average.  Then we could either expect inner-loop heartbeat at 10ms intervals, or outer-loop heartbeat at 1s intervals.  We'd need adequate safety margin on either case.  If inner-loop, we could fail a missed heartbeat more quickly, but, honestly i don't think its critical to detect the stuck watchdog with sub-second accuracy.  Maybe detecting it at minute-level accuracy is enough.  

Otoh, false alarms here will be problematic.  We'd get SEVs if our watchdog is too sensitive and kills normal programs just bc they had more works than expected.  So that also biases me to just set the heartbeat detector to a longer period - e.g. start with several minutes?  

Do we have any data on the actual time spent in the work loop for real cases?  It'd be good to validate this a bit.

Reply:
oh, and I think setting the heartbeat detector to "several minutes" and then keeping the heartbeat inside the inner-loop is also cheating.  IF you want to go that route, you'd have to convince me that it's actually "Expected" for processing one work item to sometimes take a huge amount of time.

This is not any better than multiplying "several minutes" x "a hundred work items on average" and just setting the heartbeat detector to "several hundred minutes" with an outer-loop heartbeat.

Reply:
> It is possible we have a large number of work's and each one takes some amount of time to process.

That's exactly why I am doing that. And I agree false alarm will make us look really bad. Maybe we can do it in the minute level and make the timeout configurable?

Reply:
If we set minute-level heartbeat, we don't need the inner heartbeat.

Reply:
> Do we have any data on the actual time spent in the work loop for real cases? It'd be good to validate this a bit.

We don't...

Reply:
But for NCCL hang, it's reasonable to assume, we want to kill it after couple minutes?